### PR TITLE
nodePortAccess, experimental spec override flag

### PR DESCRIPTION
**Pull request by justinsb**
This will allow us to set CIDRs for nodeport access, which in turn will
allow e2e tests that require nodeport access to pass.

Then add a feature-flagged flag to `kops create cluster` to allow
arbitrary setting of spec values; currently the only value supported is
cluster.spec.nodePortAccess

### CODE REVIEW

File: cmd/kops/create_cluster.go [hunk 1]
```diff
@@ -1039,3 +1050,22 @@ func parseCloudLabels(s string) (map[string]string, error) {
...
+		if len(kv) != 2 {
+			return fmt.Errorf("unhandled override: %q", override)
+		}
+
+		// For now we have hard-code the values we want to support; we'll get test coverage and then do this properly...
+		switch kv[0] {
+		case "cluster.spec.nodePortAccess":
+			cluster.Spec.NodePortAccess = append(cluster.Spec.NodePortAccess, kv[1])
+		default:
+			return fmt.Errorf("unhandled override: %q", override)
```
Comment (chrislovecnm) [inline cmd/kops/create_cluster.go]: I would probably not understand this error.  

How about:

> Only cluster.Spec.NodePortAccess is supported, and %s was set  More support is planned for the future.

File: pkg/apis/kops/cluster.go [hunk 2]
```diff
@@ -92,6 +92,8 @@ type ClusterSpec struct {
 	NonMasqueradeCIDR string `json:"nonMasqueradeCIDR,omitempty"`
 	// SSHAccess is a list of the CIDRs that can access SSH.
 	SSHAccess []string `json:"sshAccess,omitempty"`
+	// NodePortAccess is a list of the CIDRs that can access the node ports range (30000-32767).
+	NodePortAccess []string `json:"nodePortAccess,omitempty"`
```
Comment (chrislovecnm) [inline pkg/apis/kops/cluster.go]: nit pick, but the name does not make sense to me ...  It is an API value so we need to have it named well.  @blakebarnett / @alrs / @KashifSaadat  thoughts?  NodePortAccess sounds like a true / false. Not a list of CIDRS

NodePortSecGroupRanges?

File: pkg/apis/kops/v1alpha2/cluster.go [hunk 3]
```diff
@@ -109,9 +109,11 @@ type ClusterSpec struct {
 	NonMasqueradeCIDR string `json:"nonMasqueradeCIDR,omitempty"`
 
 	// SSHAccess determines the permitted access to SSH
-	// Currently only a single CIDR is supported (though a richer grammar could be added in future)
```
Comment (chrislovecnm) [inline pkg/apis/kops/v1alpha2/cluster.go]: It this changing in this PR?

File: cmd/kops/create_cluster.go [hunk 4]
```diff
@@ -1039,3 +1050,22 @@ func parseCloudLabels(s string) (map[string]string, error) {
 	}
 	return m, nil
 }
+
+// setOverrides sets override values in the spec
+func setOverrides(overrides []string, cluster *api.Cluster, instanceGroups []*api.InstanceGroup) error {
+	for _, override := range overrides {
+		kv := strings.SplitN(override, "=", 2)
+		if len(kv) != 2 {
+			return fmt.Errorf("unhandled override: %q", override)
```
Comment (chrislovecnm) [inline cmd/kops/create_cluster.go]: > unhandled override

Is this actually override is in the wrong format? How about

> --override must be in key=value format.  For example cluster.spec.nodePortAccess=0.0.0.0/0

File: cmd/kops/create_cluster.go [hunk 5]
```diff
@@ -284,6 +287,10 @@ func NewCmdCreateCluster(f *util.Factory, out io.Writer) *cobra.Command {
 
 	cmd.Flags().StringVar(&options.APILoadBalancerType, "api-loadbalancer-type", options.APILoadBalancerType, "Sets the API loadbalancer type to either 'public' or 'internal'")
 
+	if featureflag.SpecOverrideFlag.Enabled() {
+		cmd.Flags().StringSliceVar(&options.Overrides, "override", options.Overrides, "Directly configure values in the spec")
```
Comment (chrislovecnm) [inline cmd/kops/create_cluster.go]: nit pick.  Are these `overrides`?  Plural?

File: pkg/apis/kops/cluster.go (hunk 2 above)
Comment (justinsb) [reply pkg/apis/kops/cluster.go]: It is supposed to be a parallel to kubernetesApiAccess and sshAccess, but open to better ideas.

Comment (chrislovecnm) [reply pkg/apis/kops/cluster.go]: That makes sense.

**Comment by chrislovecnm** [comment]
@justinsb did you want to update the `fmt.Errorf` in this PR or a followup?

**Comment by justinsb** [comment]
Let's do a follow-up - I'm hoping to find a trick to make it easy to set those, and it is guarded behind a feature flag for now.

File: cmd/kops/create_cluster.go (hunk 5 above)
Comment (justinsb) [reply cmd/kops/create_cluster.go]: The idea is that you pass the flag repeatedly.  So there are many overrides, but only one per flag.

File: cmd/kops/create_cluster.go (hunk 4 above)
Comment (justinsb) [reply cmd/kops/create_cluster.go]: TBH I'm not yet sure what the overrides should look like.  They could be JSON, for example.

I figured I would guard it behind a feature flag, we can start using it to get test coverage, and we can see what "feels" right.

File: cmd/kops/create_cluster.go (hunk 1 above)
Comment (justinsb) [reply cmd/kops/create_cluster.go]: Right - hopefully this won't be a permanent state of affairs - and I certainly don't intend to hard-code every field :-)

File: pkg/apis/kops/v1alpha2/cluster.go (hunk 3 above)
Comment (justinsb) [reply pkg/apis/kops/v1alpha2/cluster.go]: No, but I spotted that this comment was wrong and I figured it wasn't _entirely_ unrelated ;-)

**Comment by chrislovecnm** [comment]
Certain irony about an enhancement that is meant to help e2e to fail when we cannot create loadbalancers, and while the enhancement is not enabled, e2e fails because of a flake.

/retest

**Comment by krzyzacy** [comment]
ref https://github.com/kubernetes/test-infra/pull/4441

**Comment by krzyzacy** [comment]
whoops - kops-k8s presubmit goes less smoothly than k8s-kops I'll need to take another look

**Comment by krzyzacy** [comment]
(that looks like some legit :snowflake: )

**Comment by justinsb** [comment]
/retest

(Fixed test, I believe)

**Comment by krzyzacy** [comment]
`ResumableUploadAbortException: 403 211744435552-compute@developer.gserviceaccount.com does not have storage.objects.create access to bucket kops-ci`

Grrr... I'm debugging with https://github.com/kubernetes/test-infra/pull/4461

**Comment by krzyzacy** [comment]
hey it works :-)

**Pull request merged**